### PR TITLE
add decode/encode instances for NullOrUndefined

### PR DESCRIPTION
**Pull request by justinwoo**

### CODE REVIEW

File: src/Data/Foreign/Class.purs [hunk 1]
```diff
@@ -89,3 +91,9 @@ instance intEncode :: Encode Int where
 
 instance arrayEncode :: Encode a => Encode (Array a) where
   encode = toForeign <<< map encode
+
+instance decodeNullOrUndefined :: (Decode a) => Decode (NullOrUndefined a) where
```
Comment (paf31) [inline src/Data/Foreign/Class.purs]: The parens here are unnecessary FYI.

File: src/Data/Foreign/NullOrUndefined.purs [hunk 2]
```diff
@@ -28,3 +28,5 @@ unNullOrUndefined (NullOrUndefined m) = m
 readNullOrUndefined :: forall a. (Foreign -> F a) -> Foreign -> F (NullOrUndefined a)
 readNullOrUndefined _ value | isNull value || isUndefined value = pure (NullOrUndefined Nothing)
 readNullOrUndefined f value = NullOrUndefined <<< Just <$> f value
+
+foreign import writeNull :: Foreign
```
Comment (paf31) [inline src/Data/Foreign/NullOrUndefined.purs]: Can we just call this `null`?

Comment (paf31) [inline src/Data/Foreign/NullOrUndefined.purs]: Also, wouldn't `undefined` be preferable, so it doesn't show up in records?

File: src/Data/Foreign/Class.purs (hunk 1 above)
Comment (justinwoo) [inline src/Data/Foreign/Class.purs]: oh, right

File: src/Data/Foreign/NullOrUndefined.purs (hunk 2 above)
Comment (justinwoo) [inline src/Data/Foreign/NullOrUndefined.purs]: ah, yeah. i ran into an issue with not being able to read `undefined` as JSON though, and the test i've pushed gives me `[null]` for `[NullOrUndefined (Nothing :: Maybe String)]`...

Comment (justinwoo) [inline src/Data/Foreign/NullOrUndefined.purs]: tests:

```purescript
testRoundTrip [NullOrUndefined (Just "test")]
testRoundTrip [NullOrUndefined (Nothing :: Maybe String)]
```

output:

```purescript
["test"]
[null]
```

looks kinda janky... maybe something i need to look into from purescript-foreign?

Comment (justinwoo) [inline src/Data/Foreign/NullOrUndefined.purs]: oh... i don't know javascript:

```
JSON.stringify([undefined])
"[null]"
```

**Comment by justinwoo** [comment]
is this ready to go?

File: src/Data/Foreign/NullOrUndefined.purs [hunk 3]
```diff
@@ -28,3 +28,5 @@ unNullOrUndefined (NullOrUndefined m) = m
 readNullOrUndefined :: forall a. (Foreign -> F a) -> Foreign -> F (NullOrUndefined a)
 readNullOrUndefined _ value | isNull value || isUndefined value = pure (NullOrUndefined Nothing)
 readNullOrUndefined f value = NullOrUndefined <<< Just <$> f value
+
+foreign import writeUndefined :: Foreign
```
Comment (paf31) [inline src/Data/Foreign/NullOrUndefined.purs]: You can use `undefined` if you escape it as `exports["undefined"] = undefined;` on the JS side, I think.

**Comment by justinwoo** [comment]
btw this is updated now with 'undefined' for the FFI. any other tweaks i should put in beforehand?

**Comment by paf31** [comment]
Thanks, and sorry for the delay!

**Pull request merged**